Avoid re-extracting DataPtrs when forwarding values between Futures

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57443 Make RRefContext get devices from RPC agent when creating OwnerRRef
* #57442 Replace DeviceIndexes with Devices in RRefs
* #56918 Make DataPtr extraction in CUDAFuture faster for Python values
* #57432 Ensure devices are preserved when forwarding between futures
* **#57433 Avoid re-extracting DataPtrs when forwarding values between Futures**

In a bunch of cases we need to "forward" between one future and another, typically because we need to convert the type of the data (e.g., from Message to PyObject). In most of these cases the DataPtrs of the value don't change, and yet the new future must re-extract them from scratch. By allowing the user to obtain the vector of extracted DataPtrs from the old future, we can allow them to "shortcut" this step.

Differential Revision: [D28118298](https://our.internmc.facebook.com/intern/diff/D28118298/)

lw committed:
Avoid re-extracting DataPtrs when forwarding values between Futures

In a bunch of cases we need to "forward" between one future and another, typically because we need to convert the type of the data (e.g., from Message to PyObject). In most of these cases the DataPtrs of the value don't change, and yet the new future must re-extract them from scratch. By allowing the user to obtain the vector of extracted DataPtrs from the old future, we can allow them to "shortcut" this step.

Differential Revision: [D28118298](https://our.internmc.facebook.com/intern/diff/D28118298/)

[ghstack-poisoned]

lw committed:
Update on "Avoid re-extracting DataPtrs when forwarding values between Futures"


In a bunch of cases we need to "forward" between one future and another, typically because we need to convert the type of the data (e.g., from Message to PyObject). In most of these cases the DataPtrs of the value don't change, and yet the new future must re-extract them from scratch. By allowing the user to obtain the vector of extracted DataPtrs from the old future, we can allow them to "shortcut" this step.

Differential Revision: [D28118298](https://our.internmc.facebook.com/intern/diff/D28118298/)

[ghstack-poisoned]

mrshenli commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -1034,7 +1034,7 @@ TORCH_API intrusive_ptr<ivalue::Future> collectAny(
         dst->setError(src->exception_ptr());
       } else {
-        dst->markCompleted(src->constValue());
+        dst->markCompleted(src->constValue(), src->dataPtrs());
```
Is it because collectAll's dst future is a join instead of child, so we cannot pass `dataPtrs` there? If so, shall we add a brief comment to explain that?

mrshenli replied:
NVM, I just saw this is already added in the next PR.

lw committed:
Update on "Avoid re-extracting DataPtrs when forwarding values between Futures"


In a bunch of cases we need to "forward" between one future and another, typically because we need to convert the type of the data (e.g., from Message to PyObject). In most of these cases the DataPtrs of the value don't change, and yet the new future must re-extract them from scratch. By allowing the user to obtain the vector of extracted DataPtrs from the old future, we can allow them to "shortcut" this step.

Differential Revision: [D28118298](https://our.internmc.facebook.com/intern/diff/D28118298/)

[ghstack-poisoned]
